### PR TITLE
Add Baremetal provider

### PR DESCRIPTION
**Pull request by mohamed-rafraf**
**What this PR does / why we need it**:

Adding a new cloud provider(bare metal) which is responsible for the provision and joining bare-metal nodes to KKP user clusters.  The bare-metal machines will be configured by using OSM in order to join the cluster. 

**Which issue(s) this PR fixes**:

Fixes #

**What type of PR is this?**

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
Adding Baremetal Provider
```

**Documentation**:

```documentation
NONE
```

### CODE REVIEW

File: pkg/apis/kubermatic/v1/helper/provider.go [hunk 1]
```diff
@@ -64,6 +64,12 @@ func ClusterCloudProviderName(spec kubermaticv1.CloudSpec) (string, error) {
...
 	}
+	if spec.Baremetal != nil {
+		clouds = append(clouds, kubermaticv1.BaremetalCloudProvider)
+	}
+	if spec.BringYourOwn != nil {
+		clouds = append(clouds, kubermaticv1.BringYourOwnCloudProvider)
+	}
 	if spec.BringYourOwn != nil {
 		clouds = append(clouds, kubermaticv1.BringYourOwnCloudProvider)
 	}
```
Comment (ahmedwaleedmalik) [inline pkg/apis/kubermatic/v1/helper/provider.go]: Remove this duplicate

**Comment by xrstf** [comment]
Before we commit to this, should it be `bareMetal` instead of `baremetal`? It's two words, isn't it? Or is baremetal a standalone word already nowadays?

**Comment by mohamed-rafraf** [comment]
> What is the difference between BYO and BareMetal?
* BYO provider, you set up nodes, configure the packages and dependencies, and join the nodes manually.
* Baremetal is used to automate the provisioning of bare-metal servers by using the Tinkerbell plugin of MC bare-metal provider and OSM to configure and join the nodes to the control plane

> Is there any chance this new provider can be tested automatically in e2e tests?

We are still exploring the possibilities to have e2e tests. Implementing e2e tests for a bare-metal provisioning engine involves various challenges and dependencies that we need to consider

**Comment by mohamed-rafraf** [comment]
> Before we commit to this, should it be `bareMetal` instead of `baremetal`? It's two words, isn't it? Or is baremetal a standalone word already nowadays?

The term `baremetal` as one word is becoming more common, also `machine-controller` already has a `baremetal` [provider](https://github.com/kubermatic/machine-controller/blob/main/pkg/providerconfig/types/types.go#L79). So I think we can go with `baremetal` as one word.

**Comment by ahmedwaleedmalik** [comment]
I have nothing against this but eventually, this will become quite confusing since now we have edge, BYOC, and baremetal providers that have more in common than different.

Anyways, how do we envision this from a user prospective? This provider doesn't require any preset/credentials or data center specific configurations?

**Comment by moadqassem** [comment]
> I have nothing against this but eventually, this will become quite confusing since now we have edge, BYOC, and baremetal providers that have more in common than different.
> 
> 
> 
> Anyways, how do we envision this from a user prospective? This provider doesn't require any preset/credentials or data center specific configurations? 
> 
> 
> 
> The problem right now is that you are working on PRs directly but there is no issue/ticket against them that we can track to see what the end goal is.

Just to shed some light on this particular project. The baremetal provider is a continuation of an old project that all the issues the tickets probably have been archived since it hasn't been touched for ages. Coming back to your questions regarding the differences. Edge as the name suggests it's mainly used for edge devices such IoT appliances and the provider doesn't assume that, an OS exists and no PXE booting will take a place since the device might be a remote host that doesn't support that process. BYOD is also a manual process where an OS exists with all the needed dependencies such as container runtime, Kubernetes binaries,etc and soon this provider will be replaced by the edge provider as those two have really no major differences. 

Coming to the bare metal provider the target is pretty different than the other two, as it focuses mainly on the provisioning of bare metal machines in private data centers, where ZTP(zero touch provisioning) approach is introduced. 

And eventually there will be an extensive documentation that @mohamed-rafraf will introduce before releasing this thing. 

Let me know if you need to get more information regarding the provider 😉

**Comment by ahmedwaleedmalik** [comment]
> > I have nothing against this but eventually, this will become quite confusing since now we have edge, BYOC, and baremetal providers that have more in common than different.
> > Anyways, how do we envision this from a user prospective? This provider doesn't require any preset/credentials or data center specific configurations?
> > The problem right now is that you are working on PRs directly but there is no issue/ticket against them that we can track to see what the end goal is.
> 
> Just to shed some light on this particular project. The baremetal provider is a continuation of an old project that all the issues the tickets probably have been archived since it hasn't been touched for ages. Coming back to your questions regarding the differences. Edge as the name suggests it's mainly used for edge devices such IoT appliances and the provider doesn't assume that, an OS exists and no PXE booting will take a place since the device might be a remote host that doesn't support that process. BYOD is also a manual process where an OS exists with all the needed dependencies such as container runtime, Kubernetes binaries,etc and soon this provider will be replaced by the edge provider as those two have really no major differences.
> 
> Coming to the bare metal provider the target is pretty different than the other two, as it focuses mainly on the provisioning of bare metal machines in private data centers, where ZTP(zero touch provisioning) approach is introduced.
> 
> And eventually there will be an extensive documentation that @mohamed-rafraf will introduce before releasing this thing.
> 
> Let me know if you need to get more information regarding the provider 😉

Thanks for the detailed explanation, @mohamed-rafraf please create a new ticket for `baremetal` provider, put the details in there, and we can track the changes through it.

**Review by ahmedwaleedmalik — APPROVED** [review]
/approve

**Comment by xrstf** [comment]
/hold

let's have at least one short discussion about this in SIG cluster.

**Comment by xrstf** [comment]
/lgtm cancel
/unhold

@mohamed-rafraf you need to rebase and re-generate the fixtures :)

**Review by ahmedwaleedmalik — APPROVED** [review]
/approve

**Pull request merged**